beluga/include/beluga/sensor/ndt_sensor_model.hpp

```diff
@@ -0,0 +1,234 @@
+  return os;
+}
+
+/// Given a number of 2D points and a resolution, constructs a vector of NDT cells. TODO(Ramiro) fill this out.
+inline std::vector<NDTCell> to_cells(const std::vector<Eigen::Vector2d>& points, const double resolution) {
+  static constexpr int kMinPointsPerCell = 5;
+
+  Eigen::Map<const Eigen::Matrix2Xd> points_view(
+      reinterpret_cast<const double*>(points.data()), 2, static_cast<int64_t>(points.size()));
+
+  std::vector<NDTCell> ret;
+  ret.reserve(static_cast<size_t>(points_view.cols()) / kMinPointsPerCell);
+
+  std::unordered_map<Eigen::Vector2i, std::vector<Eigen::Vector2d>, CellHasher> cell_grid;
+  for (const Eigen::Vector2d& col : points) {
+    cell_grid[(col / resolution).cast<int>()].emplace_back(col);
+  }
+
+  for (const auto& [cell, points_in_cell] : cell_grid) {
+    if (points_in_cell.size() < kMinPointsPerCell) {
+      continue;
+    }
+    ret.push_back(fit_points(points_in_cell));
+  }
+
+  return ret;
+}
+
+/// NDT sensor model for range finders.
+/**
+ * This class satisfies \ref SensorModelPage.
+ *
+ * \tparam SparseGridT Type representing a sparse NDT grid, as a specialization of 'beluga::SparseValueGrid'.
+ */
+template <typename SparseGridT>
+class NDTSensorModel {
+ public:
+  static_assert(std::is_same_v<SparseValueGrid<typename SparseGridT::map_type>, SparseGridT>);
+  static_assert(std::is_same_v<typename SparseGridT::mapped_type, NDTCell>);
+  /// State type of a particle.
+  using state_type = Sophus::SE2d;
+  /// Weight type of the particle.
+  using weight_type = double;
+  /// Measurement type of the sensor: a point cloud for the range finder.
+  using measurement_type = std::vector<Eigen::Vector2d>;
+  /// Map representation type.
+  using map_type = SparseGridT;
+  /// Parameter type that the constructor uses to configure the NDT sensor model.
+  using param_type = NDTModelParam;
+
+  /// Constructs a NDTSensorModel instance.
+  /**
+   * \param params Parameters to configure this instance.
+   *  See beluga::NDTModelParam for details.
+   * \param cells_data Sparse grid representing an NDT map, used for calculating importance weights for the different
+   * particles.
+   */
+  NDTSensorModel(param_type param, SparseGridT cells_data)
+      : params_{std::move(param)}, cells_data_{std::move(cells_data)} {}
+
+  /// Returns a state weighting function conditioned on 2D lidar hits.
+  /**
+   * \param points 2D lidar hit points in the reference frame of particle states.
+   * \return a state weighting function satisfying \ref StateWeightingFunctionPage
+   *  and borrowing a reference to this sensor model (and thus their lifetime are bound).
+   */
+  [[nodiscard]] auto operator()(measurement_type&& points) const {
+    return [this, cells = to_cells(points, cells_data_.resolution())](const state_type& state) -> weight_type {
+      return std::transform_reduce(
+          cells.cbegin(), cells.cend(), 1.0, std::plus{},
```

Comment:
@serraramiro1 why do we start at 1?

Reply:
I just copied this from `likelihood_sensor_model.hpp`, I figured starting from 0 might cause weird normalization / recovery behaviors.

WDYT?

Reply:
> I figured starting from 0 might cause weird normalization / recovery behaviors.

Why? It's funny, you can trace this back to 20yo AMCL code, which gives no reasons for these deviations. 

I'd rather not propagate it unless we have a good reason to.